app/lib/search/backend.dart

```diff
@@ -135,41 +136,60 @@ class SearchBackend {
   }
 
-  List<ApiDocPage> _apiDocPagesFromIndexJson(List<int> bytes) {
+  List<ApiDocPage> _apiDocPagesFromPubData(List<int> bytes) {
     if (bytes == null) return null;
     try {
-      final list = json.decode(utf8.decode(bytes));
+      final decodedMap = json.decode(utf8.decode(bytes)) as Map;
+      final pubData = new PubDartdocData.fromJson(decodedMap.cast());
+
+      final nameToKindMap = <String, String>{};
+      pubData.apiElements.forEach((e) {
+        nameToKindMap[e.name] = e.kind;
+      });
 
       final pathMap = <String, String>{};
       final symbolMap = <String, Set<String>>{};
-      for (Map map in list) {
-        final String name = map['name'];
-        final type = map['type'];
-        if (isCommonApiSymbol(name) && type != 'library') {
-          continue;
-        }
+      final docMap = <String, List<String>>{};
 
-        final String qualifiedName = map['qualifiedName'];
-        final enclosedBy = map['enclosedBy'];
-        final enclosedByType = enclosedBy is Map ? enclosedBy['type'] : null;
-        final parentLevel = enclosedByType == 'class' ? 2 : 1;
-        final String key = qualifiedName.split('.').take(parentLevel).join('.');
+      bool isTopLevel(String kind) => kind == 'library' || kind == 'class';
 
-        if (key == qualifiedName) {
-          pathMap[key] = map['href'] as String;
+      void update(String key, String name, String documentation) {
+        final set = symbolMap.putIfAbsent(key, () => new Set<String>());
+        set.addAll(name.split('.'));
+
+        documentation = documentation?.trim();
+        if (documentation != null && documentation.isNotEmpty) {
+          final list = docMap.putIfAbsent(key, () => []);
+          list.add(compactReadme(documentation));
         }
-        symbolMap.putIfAbsent(key, () => new Set()).add(name);
       }
 
+      pubData.apiElements.forEach((apiElement) {
+        if (isTopLevel(apiElement.kind)) {
+          pathMap[apiElement.name] = apiElement.href;
+          update(apiElement.name, apiElement.name, apiElement.documentation);
+        }
+
+        if (!isTopLevel(apiElement.kind) &&
+            apiElement.parent != null &&
+            isTopLevel(nameToKindMap[apiElement.parent])) {
+          update(apiElement.parent, apiElement.name, apiElement.documentation);
+        }
+      });
+
       final results = pathMap.keys.map((key) {
         final path = pathMap[key];
         final symbols = symbolMap[key].toList()..sort();
-        return new ApiDocPage(relativePath: path, symbols: symbols);
+        return new ApiDocPage(
+          relativePath: path,
+          symbols: symbols,
+          textBlocks: docMap[key],
+        );
       }).toList();
       results.sort((a, b) => a.relativePath.compareTo(b.relativePath));
       return results;
     } catch (e, st) {
-      _logger.warning('Parsing dartdoc index.json failed.', e, st);
+      _logger.warning('Parsing pub-data.json failed.', e, st);
```

Comment:
If pub-data.json is made by us and not by users, then that sounds like an error to me (we'll certainly want to know about it). What makes this case plausible? The return null can be moved into the catch body.

Maybe the caller should have this try block instead of wrapping the whole function? This function also doesn't know the data came from 'pub-data.json'.

Reply:
The parse error could happen if we change the format in a breaking way, and an older client reads the new format, or a newer client reads the old format, and they are not prepared for some kind of new structure.